### PR TITLE
Support iceberg table filter pushdown

### PR DESCRIPTION
**Pull request by beinan**
Implement a skeleton of the connector optimizer and predicate pushdown.  Quite a few new features and optimization could be implemented starting from here. 

```
== RELEASE NOTES ==

Iceberg Connector Changes
* Add Iceberg table predicate pushdown to reduce the number of files scanned
```

### CODE REVIEW

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergConnector.java [hunk 1]
```diff
@@ -87,6 +92,7 @@ public IcebergConnector(
         this.tableProperties = ImmutableList.copyOf(requireNonNull(tableProperties, "tableProperties is null"));
         this.accessControl = requireNonNull(accessControl, "accessControl is null");
         this.procedures = requireNonNull(procedures, "procedures is null");
+        this.planOptimizer = planOptimizer;
```
Comment (ChunxuTang) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergConnector.java:95]: nit: Refactoring it to use the `requireNonNull` function? Or could the optimizer be null?

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergPlanOptimizer.java [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    private final StandardFunctionResolution functionResolution;
+    private final TypeManager typeManager;
+
+    @Inject
+    IcebergPlanOptimizer(IcebergMetadataManager metadataManager, StandardFunctionResolution functionResolution, RowExpressionService rowExpressionService, TypeManager typeManager)
+    {
+        this.metadataManager = metadataManager;
+        this.functionResolution = functionResolution;
+        this.rowExpressionService = rowExpressionService;
+        this.typeManager = typeManager;
```
Comment (ChunxuTang) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergPlanOptimizer.java:59]: nit: `requiteNonNull`?

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/util/IcebergMetadataManager.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    IcebergMetadataManager(
+            IcebergConfig config,
+            IcebergResourceFactory resourceFactory,
+            IcebergTransactionManager transactionManager,
+            HdfsEnvironment hdfsEnvironment)
+    {
+        this.config = config;
+        this.resourceFactory = resourceFactory;
+        this.transactionManager = transactionManager;
+        this.hdfsEnvironment = hdfsEnvironment;
```
Comment (ChunxuTang) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/util/IcebergMetadataManager.java:49]: same here.

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/util/IcebergMetadataManager.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    public Table getColumnHandles(ConnectorSession session, ConnectorTransactionHandle transaction, SchemaTableName schemaTableName)
+    {
+        if (config.isNativeMode()) {
+            return resourceFactory.getCatalog(session).loadTable(toIcebergTableIdentifier(schemaTableName));
+        }
+        else {
+            ExtendedHiveMetastore metastore = ((IcebergMetadata) transactionManager.get(transaction)).getMetastore();
+            return getIcebergTable(metastore, hdfsEnvironment, session, schemaTableName);
+        }
+    }
```
Comment (ChunxuTang) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/util/IcebergMetadataManager.java:72]: Looks like the two functions here are exactly the same excepts for the names. Are there any specific reasons that you want them to be two functions?

**Review by ChunxuTang — COMMENTED** [review]
@beinan 
Nice work! We're having more optimizations for the presto-iceberg connector!!

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergPlanOptimizer.java [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+                    new TableHandle(handle.getConnectorId(), newTableHandle, handle.getTransaction(), handle.getLayout()),
+                    tableScan.getOutputVariables(),
+                    tableScan.getAssignments(),
+                    tableScan.getCurrentConstraint(),
+                    TupleDomain.all());
+
+            if (TRUE_CONSTANT.equals(filterPredicate)) {
+                return newTableScan;
+            }
+            else {
```
Comment (zhenxiao) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergPlanOptimizer.java:158]: remove this else? return directly

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergPlanOptimizer.java [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+            Map<String, IcebergColumnHandle> nameToColumnHandlesMapping = tableScan.getAssignments().entrySet().stream()
+                    .collect(Collectors.toMap(e -> e.getKey().getName(), e -> (IcebergColumnHandle) e.getValue()));
+
+            RowExpression filterPredicate = filter.getPredicate();
+
+            //TODO we should optimize the filter expression
+            DomainTranslator.ExtractionResult<Subfield> decomposedFilter = rowExpressionService.getDomainTranslator()
+                    .fromPredicate(session, filterPredicate, new SubfieldExtractor(functionResolution, rowExpressionService.getExpressionOptimizer(), session).toColumnExtractor());
+
+            // Only pushdown the range filters which apply to entire columns, because iceberg does not accept the filters on the subfields in nested structures.
```
Comment (zhenxiao) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergPlanOptimizer.java:131]: remove period at the end of one line comment? or we might have to add period to other one line comments

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/util/IcebergMetadataManager.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        this.transactionManager = transactionManager;
+        this.hdfsEnvironment = hdfsEnvironment;
+    }
+
+    public Table getTable(ConnectorSession session, ConnectorTransactionHandle transaction, SchemaTableName schemaTableName)
+    {
+        if (config.isNativeMode()) {
+            return resourceFactory.getCatalog(session).loadTable(toIcebergTableIdentifier(schemaTableName));
+        }
+        else {
```
Comment (zhenxiao) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/util/IcebergMetadataManager.java:57]: remove else, return directly

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/util/IcebergMetadataManager.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+            return getIcebergTable(metastore, hdfsEnvironment, session, schemaTableName);
+        }
+    }
+
+    public Table getColumnHandles(ConnectorSession session, ConnectorTransactionHandle transaction, SchemaTableName schemaTableName)
+    {
+        if (config.isNativeMode()) {
+            return resourceFactory.getCatalog(session).loadTable(toIcebergTableIdentifier(schemaTableName));
+        }
+        else {
```
Comment (zhenxiao) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/util/IcebergMetadataManager.java:68]: same for else

**Review by zhenxiao — COMMENTED** [review]
looks nice, @beinan 
a few minor issues

**Comment by kewang1024** [comment]
Can we rephrase the release note according to: https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergPlanOptimizer.java (hunk 2 above)
Comment (beinan) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergPlanOptimizer.java:59]: done

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergConnector.java (hunk 1 above)
Comment (beinan) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergConnector.java:95]: done

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergPlanOptimizer.java (hunk 6 above)
Comment (beinan) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergPlanOptimizer.java:131]: good call, done

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergPlanOptimizer.java (hunk 5 above)
Comment (beinan) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergPlanOptimizer.java:158]: done

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/util/IcebergMetadataManager.java (hunk 4 above)
Comment (beinan) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/util/IcebergMetadataManager.java:72]: my bad, this new file is for another refactoring PR, actually it's not used by this PR, I will remove this file from this PR.  Thanks

**Comment by beinan** [comment]
> Can we rephrase the release note according to: https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines

Thank you @kewang1024 @ChunxuTang @zhenxiao ! I fixed the code review issues, could you guys take a second look?

**Review by ChunxuTang — APPROVED** [review]
@beinan LGTM. Thanks for your work!

**Review by zhenxiao — APPROVED** [review]
nice work, @beinan

**Comment by zhenxiao** [comment]
hi @kewang1024 could you please take another look? If everything is good, I would merge this PR

**Comment by beinan** [comment]
Hello @kewang1024 , do you still wanna take another look on this PR?  Many thanks!

**Review by kewang1024 — APPROVED** [review]

**Comment by kewang1024** [comment]
Thanks @beinan @zhenxiao, the PR looks good to me

**Comment by maswin** [comment]
There seem to be a lot of compilation failure after this merge:
https://github.com/prestodb/presto/runs/4758611307?check_suite_focus=true

```
Error:  COMPILATION ERROR : 
Error:  /home/runner/work/presto/presto/presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergPlanOptimizer.java:[142,42] no suitable constructor found for TableScanNode(com.facebook.presto.spi.plan.PlanNodeId,com.facebook.presto.spi.TableHandle,java.util.List<com.facebook.presto.spi.relation.VariableReferenceExpression>,java.util.Map<com.facebook.presto.spi.relation.VariableReferenceExpression,com.facebook.presto.spi.ColumnHandle>,com.facebook.presto.common.predicate.TupleDomain<com.facebook.presto.spi.ColumnHandle>,com.facebook.presto.common.predicate.TupleDomain<java.lang.Object>)
    constructor com.facebook.presto.spi.plan.TableScanNode.TableScanNode(java.util.Optional<com.facebook.presto.spi.SourceLocation>,com.facebook.presto.spi.plan.PlanNodeId,com.facebook.presto.spi.TableHandle,java.util.List<com.facebook.presto.spi.relation.VariableReferenceExpression>,java.util.Map<com.facebook.presto.spi.relation.VariableReferenceExpression,com.facebook.presto.spi.ColumnHandle>) is not applicable
```

**Comment by zhenxiao** [comment]
found it. due to an api update. am pinging @beinan for a quick fix

**Comment by zhenxiao** [comment]
hot fix: https://github.com/prestodb/presto/pull/17168/files

**Pull request merged**